File: src/main/java/komu/blunt/asm/opcodes/OpJumpIfFalse.java
package komu.blunt.asm.opcodes;

import komu.blunt.asm.Label;
import komu.blunt.asm.Register;
import komu.blunt.asm.VM;
import komu.blunt.objects.TypeConstructorValue;

import static com.google.common.base.Preconditions.checkNotNull;

public final class OpJumpIfFalse extends OpCode {

    private final Register register;
    private final Label label;

    public OpJumpIfFalse(Register register, Label label) {
        this.register = checkNotNull(register);
        this.label = checkNotNull(label);
    }

    @Override
    public void execute(VM vm) {
        Object value = vm.get(register);
        if (isFalse(value))
            vm.pc = label.getAddress();
    }

    @Override
    public boolean modifies(Register register) {
        return register == Register.PC;
    }

    public static boolean isFalse(Object value) {
        return Boolean.FALSE.equals(value)
            || (value instanceof TypeConstructorValue && ((TypeConstructorValue) value).name.equals("False"));
    }

    @Override
    public String toString() {
        return String.format("(jump-if-false %s %s)", register, label);
    }
}

File: src/main/kotlin/komu/blunt/asm/opcodes/OpJumpIfFalse.kt
package komu.blunt.asm.opcodes

import komu.blunt.asm.Label
import komu.blunt.asm.Register
import komu.blunt.asm.VM
import komu.blunt.objects.TypeConstructorValue
import komu.blunt.types.ConstructorNames

class OpJumpIfFalse(private val register: Register, private val label: Label) : OpCode() {

    class object {
        // TODO: move this
        fun isFalse(value: Any?): Boolean =
            false == value || (value is TypeConstructorValue && value.name == ConstructorNames.FALSE)
    }

    override fun execute(vm: VM) {
        val value = vm[register]
        if (isFalse(value))
            vm.pc = label.address
    }

    override fun modifies(register: Register) = register == Register.PC

    override fun toString() = "(jump-if-false $register $label)"
}

